File: android/app/src/main/kotlin/com/qrferry/receiver/pipeline/ReceivePipeline.kt
package com.qrferry.receiver.pipeline

import com.google.mlkit.vision.barcode.common.Barcode
import com.qrferry.receiver.core.ContentType
import com.qrferry.receiver.core.ManifestPayload
import com.qrferry.receiver.core.Proto
import com.qrferry.receiver.core.ProtocolError
import com.qrferry.receiver.core.ReceiveSession
import com.qrferry.receiver.core.decodeFrame
import java.io.ByteArrayOutputStream
import java.io.File
import java.security.MessageDigest
import android.os.SystemClock
import java.util.zip.Inflater

/**
 * Layer 4 接收流水线 —— 对齐 src/qrferry/app/receive_pipeline.py。
 * 图像(ML Kit 多码) → decodeFrame → ReceiveSession → 重组解压校验。
 * 坏帧（CRC 失败/MAGIC 不符/解析异常）静默丢弃，绝不崩溃（协议 §11）。
 */
class ReceivePipeline {

    private val session = ReceiveSession()

    val progress: Double get() = session.progress
    val missingIndices: List<Int> get() = session.missingIndices
    val isComplete: Boolean get() = session.isComplete
    val started: Boolean get() = session.started
    val K: Int get() = session.K
    val manifest: ManifestPayload? get() = session.manifest

    // 链路诊断计数
    var seenBarcodes: Int = 0; private set   // ML Kit 返回的码总数
    var validFrames: Int = 0; private set    // 成功 decode + ingest
    var badFrames: Int = 0; private set      // CRC/解析失败被丢弃
    var firstValidFrameAtMs: Long? = null; private set

    val elapsedSeconds: Double
        get() = firstValidFrameAtMs?.let { (SystemClock.elapsedRealtime() - it).coerceAtLeast(0L) / 1000.0 } ?: 0.0

    /** 处理一帧识别到的 QR 码集合；返回是否本次触发完成。 */
    fun process(barcodes: List<Barcode>): Boolean {
        seenBarcodes += barcodes.size
        for (b in barcodes) {
            val raw = b.rawBytes ?: continue
            if (raw.size < Proto.HEADER_SIZE + Proto.CRC_SIZE) continue
            try {
                val (header, payload) = decodeFrame(raw)
                session.ingest(header, payload)
                validFrames++
                if (firstValidFrameAtMs == null) {
                    firstValidFrameAtMs = SystemClock.elapsedRealtime()
                }
                if (session.isComplete) return true
            } catch (e: ProtocolError) {
                badFrames++   // 坏帧丢弃
            }
        }
        return false
    }

    /** 完成后重组 → 解压 → SHA-256 校验，返回结果（TEXT 直接给文本；FILE 给字节+文件名，由上层落盘）。 */
    fun finalize(): Result {
        val m = session.manifest ?: error("会话未建立")
        val encoded = session.reassemble()
        val raw = inflateZlib(encoded)
        val shaHex = MessageDigest.getInstance("SHA-256").digest(raw).toHex()
        val shaOk = shaHex == m.rawSha256.toHex()
        return when (ContentType.of(m.contentType)) {
            ContentType.TEXT -> Result.Text(String(raw, Charsets.UTF_8), shaOk, shaHex)
            ContentType.FILE -> Result.File(
                filename = safeFilename(m.filename.ifBlank { "qrferry.bin" }),
                bytes = raw,
                shaOk = shaOk,
                shaHex = shaHex,
            )
            else -> Result.Text("<未知 content_type=${m.contentType}>", false, shaHex)
        }
    }

    sealed class Result {
        abstract val shaOk: Boolean
        abstract val shaHex: String
        data class Text(val text: String, override val shaOk: Boolean, override val shaHex: String) : Result()
        data class File(val filename: String, val bytes: ByteArray, override val shaOk: Boolean, override val shaHex: String) : Result()
    }

    companion object {
        /** 对齐 receive_pipeline.py: safe_filename —— basename + 过滤非法字符 + 截断 200。 */
        private fun safeFilename(name: String): String {
            val base = File(name).name
            val filtered = base.replace(Regex("[<>:\"/\\\\|?*\\x00-\\x1f]"), "_").trimEnd('.', ' ')
            val cleaned = if (filtered.isBlank()) "qrferry.bin" else filtered
            return if (cleaned.length > 200) cleaned.substring(0, 200) else cleaned
        }

        private fun ByteArray.toHex(): String =
            joinToString("") { (it.toInt() and 0xFF).toString(16).padStart(2, '0') }

        /** 解压标准 zlib 流（Python zlib.compress 产出；Inflater nowrap=false）。 */
        private fun inflateZlib(data: ByteArray): ByteArray {
            val inf = Inflater()
            inf.setInput(data)
            val out = ByteArrayOutputStream()
            val buf = ByteArray(4096)
            while (!inf.finished()) {
                val n = inf.inflate(buf)
                if (n == 0) break
                out.write(buf, 0, n)
            }
            inf.end()
            return out.toByteArray()
        }
    }
}

File: src/qrferry/app/receive_pipeline.py
"""接收流水线 —— 图像 → 解码 → 会话 → 落盘 的纯逻辑层。

与 Qt/摄像头解耦：上层（UI/控制器）负责采集帧图像传入，本模块负责
解码多码 → CRC 校验 → 驱动 ReceiveSession → 完成后解压 + SHA-256 校验 + 落盘/文本。
"""
from __future__ import annotations

import hashlib
import os
import time
import zlib
from dataclasses import dataclass

from qrferry.core.frame import Compression, ContentType, ProtocolError, decode_frame
from qrferry.core.session import ReceiveSession
from qrferry.qr.backend import CodecBackend, StandardQrBackend
from qrferry.app import session_store

__all__ = ["ReceiveResult", "ReceivePipeline", "safe_filename"]


@dataclass
class ReceiveResult:
    content_type: int
    filename: str
    data: bytes
    path: str | None       # 文件落盘绝对路径；文本传输为 None


def safe_filename(name: str) -> str:
    """剥离路径穿越与非法字符（协议 §11 安全模型）。"""
    base = os.path.basename(name.replace("\\", "/")).strip()
    if base in ("", ".", ".."):
        base = "received"
    for ch in '<>:"/\\|?*':
        base = base.replace(ch, "_")
    return base[:200]


class ReceivePipeline:
    """接收端核心：喂入图像，自动解码并累积，完成后产出结果。"""

    def __init__(self, session: ReceiveSession | None = None,
                 backend: CodecBackend | None = None, save_dir: str = ".",
                 resume_from: ReceiveSession | None = None, persist: bool = True):
        self.session = session or resume_from or ReceiveSession()
        self.backend = backend or StandardQrBackend()
        self.save_dir = save_dir
        self._finalized = False
        self.result: ReceiveResult | None = None
        self._persist = persist
        self._ingested_since_save = 0
        self.first_valid_frame_ts: float | None = None
        self.valid_frames = 0          # 有效帧数（CRC 通过并入会话）
        self.bad_frames = 0            # 丢弃帧数（CRC 失败/解析失败）
        self.missed_images = 0         # 物理码未解出的图像帧数
        if persist and resume_from is not None and resume_from.to_snapshot() is not None:
            session_store.save(self.session, save_dir)   # 恢复后立即落盘，固化种子

    @property
    def progress(self) -> float:
        return self.session.progress

    @property
    def missing_indices(self) -> list[int]:
        return self.session.missing_indices

    @property
    def drop_rate(self) -> float:
        """丢弃率 = bad_frames / (valid+bad)，链路质量指标。"""
        total = self.valid_frames + self.bad_frames
        return self.bad_frames / total if total else 0.0

    @property
    def elapsed_seconds(self) -> float:
        if self.first_valid_frame_ts is None:
            return 0.0
        return max(0.0, time.time() - self.first_valid_frame_ts)

    @property
    def is_complete(self) -> bool:
        return self.session.is_complete

    def process_image(self, image) -> int:
        """解码图像中所有 QR，有效帧（CRC 通过）入会话；返回本帧有效帧数。"""
        added = 0
        decoded = self.backend.decode(image)
        if not decoded:
            self.missed_images += 1
        for raw in decoded:
            try:
                header, payload = decode_frame(raw)
            except ProtocolError:
                self.bad_frames += 1   # CRC 失败/坏帧：丢弃但计数
                continue
            self.session.ingest(header, payload)
            added += 1
        self.valid_frames += added
        if added > 0 and self.first_valid_frame_ts is None:
            self.first_valid_frame_ts = time.time()
        if self.session.is_complete and not self._finalized:
            self._finalize()
        elif self._persist and added > 0:
            self._maybe_save(added)
        return added

    def _maybe_save(self, added: int) -> None:
        """节流持久化：每累计 16 个有效帧存一次，避免每帧 IO。"""
        self._ingested_since_save += added
        if self._ingested_since_save >= 16:
            self._ingested_since_save = 0
            session_store.save(self.session, self.save_dir)

    def _finalize(self) -> None:
        m = self.session.manifest
        encoded = self.session.reassemble()
        if m.compression == Compression.NONE:
            data = encoded
        elif m.compression == Compression.ZLIB:
            data = zlib.decompress(encoded)
        else:
            raise ValueError(f"暂不支持的压缩类型: {m.compression}")
        if hashlib.sha256(data).digest() != m.raw_sha256:
            raise ValueError("SHA-256 校验失败：数据损坏")
        if m.content_type == ContentType.TEXT:
            self.result = ReceiveResult(m.content_type, "", data, None)
        else:
            os.makedirs(self.save_dir, exist_ok=True)
            path = os.path.join(self.save_dir, safe_filename(m.filename))
            with open(path, "wb") as f:
                f.write(data)
            self.result = ReceiveResult(m.content_type, m.filename, data, path)
        self._finalized = True
        if self._persist:
            session_store.clear(self.save_dir)   # 传输完成，清理续传快照
